File: services/analyzer/analyzers/color.py
"""Color analyzer: dominant palette + WCAG contrast between text and its
local background, for every OCR-detected text region."""

from __future__ import annotations

import cv2
import numpy as np

from models import Finding, Measured
from shared import TextRegion, new_id, wcag_contrast

WCAG_AA_NORMAL = 4.5


def _dominant_colors(patch: np.ndarray, k: int = 2) -> np.ndarray:
    pixels = patch.reshape(-1, 3).astype(np.float32)
    if len(pixels) < k:
        return np.array([pixels.mean(axis=0)] * k)
    criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 20, 0.5)
    _, labels, centers = cv2.kmeans(pixels, k, None, criteria, 3, cv2.KMEANS_PP_CENTERS)
    counts = np.bincount(labels.flatten(), minlength=k)
    order = np.argsort(-counts)
    return centers[order]


def analyze(bgr: np.ndarray, text_regions: list[TextRegion]) -> list[Finding]:
    findings: list[Finding] = []
    h, w = bgr.shape[:2]

    for region in text_regions:
        x, y, rw, rh = [int(v) for v in region.bbox]
        x, y = max(0, x), max(0, y)
        rw, rh = max(1, min(rw, w - x)), max(1, min(rh, h - y))
        pad = max(4, rh // 2)
        y0, y1 = max(0, y - pad), min(h, y + rh + pad)
        x0, x1 = max(0, x - pad), min(w, x + rw + pad)
        patch = bgr[y0:y1, x0:x1]
        if patch.size == 0:
            continue

        # Two-cluster split approximates "text ink color" vs "local
        # background color" without needing per-pixel text segmentation.
        centers_bgr = _dominant_colors(patch, k=2)
        centers_rgb = centers_bgr[:, ::-1]
        contrast = wcag_contrast(centers_rgb[0], centers_rgb[1])

        if contrast < WCAG_AA_NORMAL:
            findings.append(
                Finding(
                    id=new_id(),
                    dimension="color",
                    severity="critical" if contrast < 3.0 else "major",
                    bbox=(x, y, rw, rh),
                    measured=Measured(value=round(float(contrast), 2), expected=(WCAG_AA_NORMAL, 21.0), unit=":1"),
                )
            )

    return findings

File: src/lib/measure/analyzers/color.ts
import type { TrackAFinding } from "@/lib/critique/types";
import { kmeans2, wcagContrast } from "../ops";
import { newFindingId } from "../id";
import type { TextLine } from "../text-lines";

/**
 * Color analyzer: dominant palette + WCAG contrast between text and its
 * local background, for every detected text line. Port of
 * `services/analyzer/analyzers/color.py`.
 */
const WCAG_AA_NORMAL = 4.5;

export function analyzeColor(
  rgb: Uint8ClampedArray,
  width: number,
  height: number,
  textLines: TextLine[],
): TrackAFinding[] {
  const findings: TrackAFinding[] = [];

  for (const region of textLines) {
    let [x, y, rw, rh] = region.bbox.map((v) => Math.round(v));
    x = Math.max(0, x);
    y = Math.max(0, y);
    rw = Math.max(1, Math.min(rw, width - x));
    rh = Math.max(1, Math.min(rh, height - y));
    const pad = Math.max(4, Math.floor(rh / 2));
    const y0 = Math.max(0, y - pad);
    const y1 = Math.min(height, y + rh + pad);
    const x0 = Math.max(0, x - pad);
    const x1 = Math.min(width, x + rw + pad);
    if (y1 <= y0 || x1 <= x0) continue;

    const patchW = x1 - x0;
    const patchH = y1 - y0;
    const pixels = new Float32Array(patchW * patchH * 3);
    let n = 0;
    for (let py = y0; py < y1; py++) {
      for (let px = x0; px < x1; px++) {
        const src = (py * width + px) * 3;
        pixels[n++] = rgb[src];
        pixels[n++] = rgb[src + 1];
        pixels[n++] = rgb[src + 2];
      }
    }

    const { centers } = kmeans2(pixels, 2);
    const contrast = wcagContrast(centers[0], centers[1]);

    if (contrast < WCAG_AA_NORMAL) {
      findings.push({
        id: newFindingId(),
        dimension: "color",
        severity: contrast < 3.0 ? "critical" : "major",
        bbox: [x, y, rw, rh],
        measured: {
          value: Math.round(contrast * 100) / 100,
          expected: [WCAG_AA_NORMAL, 21.0],
          unit: ":1",
        },
      });
    }
  }

  return findings;
}
